src/core/lib/transport/metadata.h

```diff
@@ -109,7 +109,11 @@ struct grpc_mdelem {
                               (uintptr_t)GRPC_MDELEM_STORAGE_INTERNED_BIT))
 
 /* Unrefs the slices. */
-grpc_mdelem grpc_mdelem_from_slices(grpc_slice key, grpc_slice value);
+grpc_mdelem grpc_mdelem_from_slices(const grpc_slice& key,
+                                    const grpc_slice& value);
+/* Does not unref the slices. */
+grpc_mdelem grpc_mdelem_from_slices_no_unref(const grpc_slice& key,
```

Comment:
Instead of having two versions of this method, could we perhaps convert the `grpc_slice` API to C++ and change these methods to take the slices as `RefCountedPtr<>`s?  That way, the semantic will be clear from the API.

Reply:
Thank you @markdroth but `grpc_slice` is ref-counted intrusively (internally). I'm not completely sure if that's something we can do with `RefCountedPtr<>`. Moving it to C++ API is a great long term plan, but is perhaps would take a long time.

If you have strong preference for having one version of this function, an easier option is to remove the unref from `grpc_mdelem_from_slice`, and calling `unref` explicitly in the call site. Would that be something you prefer over this patch?

FWIW, we already ahve `grpc_slice_sub_no_ref` and `grpc_slice_sub`.

Reply:
> Thank you @markdroth but `grpc_slice` is ref-counted intrusively (internally). I'm not completely sure if that's something we can do with `RefCountedPtr<>`. Moving it to C++ API is a great long term plan, but is perhaps would take a long time.

We have an `InternallyRefCounted` base class for internally ref-counted objects.  I haven't looked at the details of the `grpc_slice` implementation, but I'd be surprised if we couldn't find a way to make this work with our existing C++ abstractions.

But, I do understand that this would be a more invasive change.  If you don't want to do it now, that's fine.

> If you have strong preference for having one version of this function, an easier option is to remove the unref from `grpc_mdelem_from_slice`, and calling `unref` explicitly in the call site. Would that be something you prefer over this patch?
> 
> FWIW, we already ahve `grpc_slice_sub_no_ref` and `grpc_slice_sub`.

My real concern here is that I want the semantics of this API to be clear and easy to understand.  I haven't looked at the code in detail, but the fact that we sometimes want the mdelem to take a ref to the slices and sometimes don't seems like a code smell.  I suspect that there's a deeper design problem here, and adding a second entry point may just be papering over the problem instead of fixing it.

I suggest looking at all of the other callers of `grpc_mdelem_from_slices()` to see what they're doing.  If we can come up with a unified approach that works for all callers, that seems superior to this two-entry-point approach.

Reply:
Thank you @markdroth! I agree with all your statements and tried to make it a RAII class, so that we can have automatic `unref()`'s when it's needed, but I hit a wall. The problem is `grpc_slice` is unfortunately part of the C API in `include/grpc/grpc.h`. So, that effort failed.

I think the cleanest way would be really removing `unref()` calls from `grpc_mdelem_from_slices()`, and moving `unref()` calls to call sites when it's necessary. In other words, `grpc_mdelem_from_slices()` doesn't really have any business `unref()`ing the input slice. It should simply created a mdelem from the given slices and that's all. WDYT?

Reply:
> Thank you @markdroth! I agree with all your statements and tried to make it a RAII class, so that we can have automatic `unref()`'s when it's needed, but I hit a wall. The problem is `grpc_slice` is unfortunately part of the C API in `include/grpc/grpc.h`. So, that effort failed.

Right, we can't change the public C slice API.  But we could make a C++ class to use internally, and make it accessible externally via the existing public C API.  In other words, the public C API would basically be a thin wrapper around the C++ implementation.

Again, though, as I said above, if you don't want to do this now, I am fine with deferring this for work for now.

> I think the cleanest way would be really removing `unref()` calls from `grpc_mdelem_from_slices()`, and moving `unref()` calls to call sites when it's necessary. In other words, `grpc_mdelem_from_slices()` doesn't really have any business `unref()`ing the input slice. It should simply created a mdelem from the given slices and that's all. WDYT?

If it turns out that that approach seems to better match the expectations of most callers, that's fine.  I don't have time to look at the callers to see if that's true, but it seems reasonable to me if it is.

I'd say give it a try and see how it looks.

Reply:
> Again, though, as I said above, if you don't want to do this now, I am fine with deferring this for work for now.

Ack. I tried this and it's a quite intensive change. I'm not very familiar with the gRPC process, but it might be worth adding that in issues so that we don't forget about.

> I'd say give it a try and see how it looks.

Thank you @markdroth! While giving that a try, it occurred to me that `grpc_mdelem_create()` already exists and instead of adding a new variant `grpc_mdelem_from_slices_no_unref()`, I could simply use `grpc_mdelem_create()`.

It looks cleaner IMHO, and it doesn't introduce any new function to the mdelem API. So, in a nutshell, `grpc_mdelem_from_slices()` are now replaced with `grpc_mdelem_create()` when we don't need to call `unref()`.